content/blog/2023/contain-root.md

```diff
@@ -0,0 +1,517 @@
+
+> RESOLVED: deprecate any existing use of body propagation
+
+Body propagation was a mistake.
+We're stuck with it now,
+but we don't have to protect it moving forward.
+
+## Containment and propagation (according to spec)
+
+[CSS Containment Module Level 2](https://drafts.csswg.org/css-contain-1/#contain-property):
+does provide some caveats
+around root/body propagation and containment:
+
+> When any containments are active
+> on either the HTML `<html>` or `<body>` elements,
+> propagation of properties from the `<body>` element
+> to the initial containing block, the viewport, or the canvas background,
+> is disabled. Notably, this affects:
+>
+> - `writing-mode`, `direction`, and `text-orientation`
+>   (see CSS Writing Modes 3 § 8 The Principal Writing Mode)
+> - `overflow` and its longhands
+>   (see CSS Overflow 3 § 3.3 Overflow Viewport Propagation)
+> - `background` and its longhands
+>   (see CSS Backgrounds 3 § 2.11.2 The Canvas Background and the HTML `<body>` Element)
+
+(Maybe this is the complete list?!)
+
+And then we get a clarification:
+
+> NOTE: Propagation to the initial containing block,
+> the viewport, or the canvas background,
+> of properties set on the html element itself is unaffected.
+
+The logic,
+as I understood it
+(and taught it at these recent conferences)
+_goeth thusly_:
+
+- Because containment needs to stop the flow of information outward,
+  it must cut off any propagation from the body to the viewport.
+- That's ok, we don't like body propagation anyway.
+- But we can't simply remove a decades-old default.
+- So _browsers_ can't apply containment on root,
+  but _authors_ [can opt-in](https://github.com/w3c/csswg-drafts/issues/6178)
+  to that behavior.
+- As long as they don't rely on body propagation, they'll be fine!
+
+I think that logic makes good sense,
+and my expectations match
+the text of the spec as I read it.
+But clearly all the browsers implemented something else.
+
+## What browsers implemented
+
+I am not a browser engineer,
+but I've been trying to parse out
+how browsers got a different answer than I did.
+
+It's not clear to me
+that browsers are 'wrong'
+in the default case,
+when the root element has default `visible` overflow.
+In that case,
+I think the spec should be clarified.
+It says to propagate from body,
+but it doesn't say what to do
+if containment breaks that propagation.
+I would _expect_ that the viewport
+should also default to `visible` overflow,
+which it should then treat as `auto`.
+But that's not clearly stated.
+
+In any case,
+we should very clearly be able to set our own
+non-default overflow on root,
+and have that value propagate to the viewport
+with or without containment.
+It works fine without:
+
+{{ embed.codepen(
+  id='jOQmzzG',
+  title='Root overflow propagates fine',
+  user='miriamsuzanne'
+) }}
+
+But when we add containment,
+our non-default overflow no longer propagates.
+This is very clearly a browser bug:
```

Comment:
Reconsider phrasing -- three uses of "clearly" or "very clearly" in quick succession.